src/core/qgsexpression.cpp

```diff
@@ -672,15 +672,28 @@ static QVariant fcnAggregate( const QVariantList& values, const QgsExpressionCon
     parameters.delimiter = value.toString();
   }
 
-  QString cacheKey = QStringLiteral( "aggfcn:%1:%2:%3:%4" ).arg( vl->id(), QString::number( static_cast< int >( aggregate ) ), subExpression, parameters.filter );
-  if ( context && context->hasCachedValue( cacheKey ) )
-    return context->cachedValue( cacheKey );
-
   QVariant result;
   if ( context )
   {
+    QString cacheKey = QStringLiteral( "aggfcn:%1:%2:%3:%4" ).arg( vl->id(), QString::number( aggregate ), subExpression, parameters.filter );
+
+    QgsExpression subExp( subExpression );
+    if ( subExp.referencedVariables().contains( "parent" ) || subExp.referencedVariables().contains( QString() ) )
+    {
+      cacheKey += ':' + qHash( context->feature() );
+    }
+
+    if ( context && context->hasCachedValue( cacheKey ) )
+      return context->cachedValue( cacheKey );
+
     QgsExpressionContext subContext( *context );
+    QgsExpressionContextScope* subScope = new QgsExpressionContextScope();
+    subScope->setVariable( "parent", context->feature() );
+    subContext.appendScope( subScope );
     result = vl->aggregate( aggregate, subExpression, parameters, &subContext, &ok );
+
+    if ( ok )
```

Comment:
The value should be cached even if it's not ok, as otherwise we'll be running all this code multiple times and always getting the invalid result. Better to cache it and shortcut the checks.